File: Yow.CoD.Finance.Domain/Services/TakePaymentCommandHandler.cs
using System.Linq;
using System.Threading.Tasks;
using Yow.CoD.Finance.Domain.Contracts;
using Yow.CoD.Finance.Domain.Model;

namespace Yow.CoD.Finance.Domain.Services
{
    public class TakePaymentCommandHandler : IHandler<TakePaymentCommand, TransactionReceipt>
    {
        private readonly IRepository<Loan> _repository;

        public TakePaymentCommandHandler(IRepository<Loan> repository)
        {
            _repository = repository;
        }

        public async Task<TransactionReceipt> Handle(TakePaymentCommand command)
        {
            Loan loan = await _repository.Get(command.AggregateId);
            loan.TakePayment(command);
            var paymentTakenEvent = loan.GetUncommitedEvents().OfType<PaymentTakenEvent>().Single();
            var receipt = new TransactionReceipt(loan.Id, loan.Version, paymentTakenEvent.TransactionId);
            await _repository.Save(loan);
            return receipt;
        }
    }
}

File: typescript/packages/domain/src/services/take-payment-command-handler.ts
import { TakePaymentCommand } from "../contracts/take-payment-command.js";
import { TransactionReceipt } from "../contracts/transaction-receipt.js";
import { PaymentTakenEvent } from "../contracts/payment-taken-event.js";
import { Loan } from "../model/loan.js";
import { Handler } from "./handler.js";
import { Repository } from "./repository.js";

export class TakePaymentCommandHandler
  implements Handler<TakePaymentCommand, TransactionReceipt>
{
  readonly #repository: Repository<Loan>;

  constructor(repository: Repository<Loan>) {
    this.#repository = repository;
  }

  async handle(command: TakePaymentCommand): Promise<TransactionReceipt> {
    const loan = await this.#repository.get(command.aggregateId);
    loan.takePayment(command);

    const paymentEvent = loan
      .getUncommittedEvents()
      .find(
        (e): e is PaymentTakenEvent => e.eventType === "PaymentTakenEvent",
      );
    if (!paymentEvent) {
      throw new Error("PaymentTakenEvent not found after takePayment");
    }
    const receipt = new TransactionReceipt(
      loan.id,
      loan.version,
      paymentEvent.transactionId,
    );

    await this.#repository.save(loan);
    return receipt;
  }
}
